R/pkg_ref_cache_vignettes.R

```diff
@@ -0,0 +1,72 @@
+#' Cache a list of Vignettes files from a package reference
+#'
+#' @inheritParams pkg_ref_cache
+#' @family package reference cache
+#'
+pkg_ref_cache.vignettes <- function(x, name, ...) {
+  UseMethod("pkg_ref_cache.vignettes")
+}
+
+
+
+pkg_ref_cache.vignettes.pkg_remote <- function(x, name, ...) {
+  vignettes_from_url(x$web_url)
+}
+
+
+pkg_ref_cache.vignettes.pkg_install <- function(x, name, ...) {
+  vignettes_from_dir(system.file(package = x$name))
+}
+
+
+
+pkg_ref_cache.vignettes.pkg_source <- function(x, name, ...) {
+  vignettes_from_dir(x$path)
+}
+
+
+
+#' Build a list of Vignettes files discovered within a given directory
+#'
+#' @param path a package directory path expected to contain Vignettes files
+#'
+#' @return a list of parsed Vignettes files
+#'
+vignettes_from_dir <- function(path) {
+  # accommodate unique vignettes files
+
+  folder <- c(source = "/vignettes", boundle = "/inst/doc", binary = "/doc")
+  files <- unlist(lapply(paste0(path, folder), list.files, full.names = TRUE))
+
+  if (!length(files)) return(data.frame())
+
+  file_path = unique(tools::file_path_sans_ext(files))
+  filename = basename(file_path)
+
+  res <- data.frame(filename = filename, path = file_path)
+  res <- res[tolower(res$filename) != tolower("index"), ]
+
+  res
+}
+
+#' Build a list of Vignettes files discovered within a package website
+#'
+#' @param path a package directory path expected to contain Vignettes files
+#'
+#' @return a data frame of Vignettes files
+#'
+vignettes_from_url <- function(url) {
+
+  file_path <- xml2::read_html(url) %>%
+    rvest::html_nodes("a") %>%
+    rvest::html_attr("href") %>%
+    grep("vignettes", x = ., value = TRUE)
```

Comment:
My preference would be to avoid using `dplyr` (and possibly `rvest`) internally as much as possible. These operations are pretty routine and should be quite manageable in base R. 

`xml2` provides some XPath based search that should be able to pull this information. It's not nearly as clean, but it would mitigate our dependencies. For an example of how to pull a link, check out the function `pkg_ref_cache.news_urls.pkg_cran_remote`. We're _almost_ to the point that we can cut out `dplyr` as a dependency, so I'd like to minimize relying on it where possible. Our only `rvest` code is some legacy stuff that's all been commented out at this point. 

If you'd prefer, I'd be fine pulling it in as is and filing another issue to cut down on the dependencies.

Reply:
agreed and updated